File: benchmarks/duplicates/bench_golomb.cpp
/*******************************************************************************
 * benchmarks/duplicates/bench_golomb.cpp
 *
 * Part of Project Thrill - http://project-thrill.org
 *
 * Copyright (C) 2015 Alexander Noe <aleexnoe@gmail.com>
 *
 * All rights reserved. Published under the BSD-2 license in the LICENSE file.
 ******************************************************************************/

#include <thrill/common/cmdline_parser.hpp>
#include <thrill/common/math.hpp>
#include <thrill/common/stats_timer.hpp>
#include <thrill/core/dynamic_bitset.hpp>

#include <cmath>
#include <random>

using namespace thrill; // NOLINT

int main(int argc, char* argv[]) {

    size_t golomb_param = 5;
    double fpr_parameter = 8;
    size_t num_elements = 1;
    size_t average_distance = 10;

    common::CmdlineParser clp;

    clp.AddSizeT('b', "golomb_param", golomb_param,
                 "Set Golomb Parameter, default: 5");

    clp.AddDouble('f', "fpr_param", fpr_parameter,
                  "Set the False Positive Rate Parameter (FPR: 1/param), default: 8");

    clp.AddSizeT('n', "elements", num_elements,
                 "Set the number of elements");

    clp.AddSizeT('d', "avg_dist", average_distance,
                 "Average distance between numbers, default: 10");

    if (!clp.Process(argc, argv))
        return -1;

    size_t space_bound =
        num_elements * (2 + common::IntegerLog2Ceil((size_t)fpr_parameter));

    core::DynamicBitset<size_t> golomb_code(space_bound, false, golomb_param);

    std::default_random_engine generator(std::random_device { } ());
    std::uniform_int_distribution<int> distribution(1, (2 * average_distance) - 1);

    common::StatsTimerStart timer;

    for (size_t i = 0; i < num_elements; ++i) {
        golomb_code.golomb_in(distribution(generator));
    }

    timer.Stop();

    std::cout
        << "RESULT"
        << " benchmark=golomb"
        << " time=" << timer.Milliseconds()
        << " bitsize=" << golomb_code.bit_size()
        << " elements=" << num_elements
        << " average_distance=" << average_distance
        << " fpr_parameter=" << fpr_parameter
        << " golomb_parameter=" << golomb_param
        << std::endl;
}

/******************************************************************************/

File: benchmarks/duplicates/bench_golomb.py
#!/usr/bin/env python2
##########################################################################
# benchmarks/duplicates/bench_golomb.py
#
# Part of Project Thrill - http://project-thrill.org
#
#
# All rights reserved. Published under the BSD-2 license in the LICENSE file.
##########################################################################

import os
import sys
import subprocess
import random
#import numpy

num_runs = 5

##### bench golomb coder

result_dir = "./bench_golomb"

param = [4,8,16,32]
bs = [5,6,7,8,9,10]
distance = [10,20,40,80]
amounts= [14,16,18,20,22]

if not os.path.exists(result_dir): os.makedirs(result_dir)

# Insert pow(2,n) random integer elements into hashtable with specific number of workers and keyspace. Perform 5 times and print median to file.
for p in param:
    for b in bs:
        for d in distance:
            for amount in amounts:
                times = []
                for _ in range(num_runs):
                    a = ['../../build/benchmarks/duplicates_bench_golomb',
                             '-n', str(pow(2,amount)), '-b', str(b),
                             '-d', str(d), '-f', str(p)]
                    print(a)
                    process = subprocess.Popen(a, stdout=subprocess.PIPE)
                    process.wait()
                    result = process.communicate()[0].split()
                    print(result)
                    assert result[0] == "RESULT"
                    times.append(int(result[2].split("=")[1]))

                time_avg = "time_avg=" + str(sum(times) / num_runs)
                print(result[0] + " " + result[1] + " " +
                    time_avg + " " + result[3] + " " + result[4] + " " +
                    result[5] + " " + result[6] + " " + result[7])

##########################################################################
